feat(compiler-cli): Adds diagnostic for misconfigured @defer triggers

SkyZeroZx opened the pull request:
Introduces a new diagnostic to detect unreachable or redundant trigger configurations in `@defer` blocks, improving error reporting and developer feedback for template misconfigurations.

## Show warning 
Case 

```html
@defer (on immediate; prefetch on idle) {
       <my-cmp />
} 
```

Message

```
 The '@defer on immediate' trigger will run before any prefetches
```
Case  

```html
@deferon timer(100ms); prefetch on timer(3000ms) {
       <my-cmp />
} 
```

Message

```
Prefetch 'timer(3000ms)' is not earlier than the main 'timer(100ms)' and will not run before rendering. Lower the prefetch delay or remove the prefetch
```


```html
@defer (on viewport; prefetch on viewport) {
       <my-cmp />
 } 
```

Message

```
Prefetch 'viewport' matches the main trigger and provides no benefit. Remove the prefetch modifier.
```
 




## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number:  https://github.com/angular/angular/issues/52739


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

JeanMeche commented on packages/compiler-cli/src/ngtsc/typecheck/extended/checks/defer_trigger_misconfiguration/index.ts:
```diff
@@ -0,0 +1,144 @@
+    >;
+
+    // Gather actual trigger objects for mains and prefetch (only defined ones)
+    const mains = mainKeys
+      .map((k) => (node.triggers as Record<string, TmplAstDeferredTrigger | undefined>)[k])
+      .filter((t): t is TmplAstDeferredTrigger => t !== undefined && t !== null);
+
+    const prefetches = prefetchKeys
+      .map((k) => (node.prefetchTriggers as Record<string, TmplAstDeferredTrigger | undefined>)[k])
+      .filter((t): t is TmplAstDeferredTrigger => t !== undefined && t !== null);
+
+    const diags: NgTemplateDiagnostic<ErrorCode.DEFER_TRIGGER_MISCONFIGURATION>[] = [];
```
nit:
```suggestion
    const mains = mainKeys
      .map((k) => node.triggers[k])
      .filter((t): t is TmplAstDeferredTrigger => t !== undefined && t !== null);

    const prefetches = prefetchKeys
      .map((k) => node.prefetchTriggers[k])
      .filter((t): t is TmplAstDeferredTrigger => t !== undefined && t !== null);

    const diags: NgTemplateDiagnostic<ErrorCode.DEFER_TRIGGER_MISCONFIGURATION>[] = [];
```
Those 2 type assertions don't look necessary

JeanMeche commented on packages/compiler-cli/src/ngtsc/typecheck/extended/checks/defer_trigger_misconfiguration/index.ts:
```diff
@@ -0,0 +1,144 @@
+          const mainDelay = main.delay;
+          const preDelay = pre.delay;
+          if (preDelay >= mainDelay) {
+            const msg = `Prefetch 'timer(${preDelay}ms)' is not earlier than the main 'timer(${mainDelay}ms)' and will not run before rendering. Lower the prefetch delay or remove the prefetch.`;
```
```suggestion
            const msg = `The Prefetch 'timer(${preDelay}ms)' is not scheduled before the main 'timer(${mainDelay}ms)', so it won’t run prior to rendering. Lower the prefetch delay or remove it.`;
```

SkyZeroZx replied:
I just updated all the observations

SkyZeroZx commented:
> Curious if it's a real use case, I can't find something like this internally or externally? https://grep.app/search?regexp=true&q=%40defer.*imm.*prefetch

I haven't found any similar cases publicly, but there are similar rules in EsLint, such as not having the same trigger as "main" and "prefetch".
I think that instead of usually relying on an external library, it would be possible for the framework itself to give us a warning.

https://github.com/angular-eslint/angular-eslint/blob/main/packages/eslint-plugin-template/docs/rules/no-call-expression.md#-invalid-code-9

For the specific case of a main timer being less than a prefetch timer, I've seen it in some private codebases.

SkyZeroZx commented on packages/compiler-cli/src/ngtsc/diagnostics/src/error_code.ts:
```diff
@@ -423,6 +423,13 @@ export enum ErrorCode {
+   * Examples: multiple main triggers, 'on immediate' together with other mains or any prefetch,
+   * prefetch timer delay that is not earlier than the main timer, or an identical prefetch
+   */
+  DEFER_TRIGGER_MISCONFIGURATION = 8021,
```
Updated

kirjs commented on adev/src/content/reference/extended-diagnostics/NG8021.md:
````diff
@@ -0,0 +1,125 @@
+Remove redundant prefetch
+
+```typescript
+@Component({
+  template: `
+    <button #loadBtn>Load</button>
+    @defer (on interaction(loadBtn)) {
+      <large-component />
+    }
+  `
+})
````
I kinda feel like it would be valuable to have `good` and `bad` examples, 

like 

Don't to this:
```typescript
```typescript
@Component({
  template: `
    <button #loadBtn>Load</button>
    @defer (on interaction(loadBtn) prefetch on timer(500ms)) {
      <large-component />
    }
  `
})
```
do this:
```typescript
@Component({
  template: `
    <button #loadBtn>Load</button>
    @defer (on interaction(loadBtn)) {
      <large-component />
    }
  `
})
```

SkyZeroZx committed:
feat(compiler-cli): Adds diagnostic for misconfigured @defer triggers

Warns when @defer blocks define unreachable or redundant triggers, such as multiple main triggers, ineffective prefetches, or timer delays not scheduled before rendering.

JeanMeche reviewed (approved):
reviewed-for: public-api
